datastore/complete_post_object.go

```diff
@@ -0,0 +1,96 @@
+package datastore
+
+import (
+	"fmt"
+
+	nex "github.com/PretendoNetwork/nex-go"
+	common_globals "github.com/PretendoNetwork/nex-protocols-common-go/globals"
+	datastore "github.com/PretendoNetwork/nex-protocols-go/datastore"
+	datastore_types "github.com/PretendoNetwork/nex-protocols-go/datastore/types"
+)
+
+func completePostObject(err error, client *nex.Client, callID uint32, param *datastore_types.DataStoreCompletePostParam) uint32 {
+	if commonDataStoreProtocol.MinIOClient == nil {
+		common_globals.Logger.Warning("MinIOClient not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.getObjectSizeByDataIDHandler == nil {
+		common_globals.Logger.Warning("GetObjectSizeByDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.updateObjectUploadCompletedByDataIDHandler == nil {
+		common_globals.Logger.Warning("UpdateObjectUploadCompletedByDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.deleteObjectByDataIDHandler == nil {
+		common_globals.Logger.Warning("DeleteObjectByDataIDHandler not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return nex.Errors.DataStore.Unknown
+	}
+
+	if param.IsSuccess {
+		bucket := commonDataStoreProtocol.s3Bucket
+		key := fmt.Sprintf("%s/%d.bin", commonDataStoreProtocol.s3DataKeyBase, param.DataID)
+
+		objectSizeS3, err := commonDataStoreProtocol.S3ObjectSize(bucket, key)
+		if err != nil {
+			common_globals.Logger.Error(err.Error())
+			return nex.Errors.DataStore.NotFound
+		}
+
+		objectSizeDB, errCode := commonDataStoreProtocol.getObjectSizeByDataIDHandler(param.DataID)
+		if errCode != 0 {
+			return errCode
+		}
+
+		if objectSizeS3 != uint64(objectSizeDB) {
+			common_globals.Logger.Errorf("Object with DataID %d did not upload correctly! Mismatched sizes", param.DataID)
+			// TODO - Is this a good error?
+			return nex.Errors.DataStore.Unknown
+		}
+
+		errCode = commonDataStoreProtocol.updateObjectUploadCompletedByDataIDHandler(param.DataID, true)
+		if errCode != 0 {
+			return errCode
+		}
+	} else {
+		errCode := commonDataStoreProtocol.deleteObjectByDataIDHandler(param.DataID)
+		if errCode != 0 {
+			return errCode
+		}
```

Comment:
We have to check if the data ID exists on the S3 server before deleting, just in case

Reply:
My intention with `deleteObjectByDataIDHandler` is to remove it from the database, not touching S3 since presumably `IsSuccess` would only be false if the upload failed right? So the file wouldn't be in S3 anyway?

I can still add an additional check for that though if you think it's an issue

Reply:
Is this handler used anywhere else? Since this is an special case, the server implementation should know that the deletion is because of failed upload

Reply:
Right now this is only used here in common but there are other places where it will eventually be used. Such as in SMM, which adds methods `PrepareAttachFile` and `CompleteAttachFile`, so this will be reused there. Also there's `DeleteObject` and `DeleteObjects`, though those come with a password so they'll likely get their own deletion functions to handle that. Also the function could be reused in the future in the gRPC API I'm going to eventually add

Reply:
Well actually, after thinking more on it, we could just delete the data that is uploaded to S3 aswell. My concern was that the data is uploaded to S3 but the client (or custom client) reported failed upload

We should also check for permissions and if the database doesn't report the upload as complete, but I think that can be done on the delete handler (at least the permissions part)

Reply:
> Well actually, after thinking more on it, we could just delete the data that is uploaded to S3 aswell. My concern was that the data is uploaded to S3 but the client (or custom client) reported failed upload

I assumed this to be what you meant originally. So I was clarifying that the intention of `DeleteObjectByDataID` is to remove the object from the database, but not S3 in this function specifically

Most of these helper methods are just me replacing database calls from the SMM implementation with a helper method instead, so that's typically the context they work under unless otherwise specified (such as `S3ObjectSize` which pulls the objects size from S3)

So the question becomes, do we add an additional check and helper method named something like `S3RemoveObject` and call that as well, or do we just keep `DeleteObjectByDataID` and assume that the server implementing that function will do it's own S3 clean up

> We should also check for permissions and if the database doesn't report the upload as complete, but I think that can be done on the delete handler (at least the permissions part)

Unless I'm misunderstanding the suggestion, this is an implementation detail really and is how things are already implemented by SMM. For example in the SMM server the `GetObjectInfoByDataID` is implemented like so

```go
func GetObjectInfoByDataID(dataID uint64) (*datastore_types.DataStoreMetaInfo, uint32) {
	errCode := IsObjectAvailable(dataID)
	if errCode != 0 {
		return nil, errCode
	}

	// The database query handling
}
```

Where `IsObjectAvailable` is implemented as

```go
func IsObjectAvailable(dataID uint64) uint32 {
	var underReview bool

	err := database.Postgres.QueryRow(`SELECT under_review FROM datastore.objects WHERE data_id=$1 AND upload_completed=TRUE AND deleted=FALSE`, dataID).Scan(&underReview)
	if err != nil {
		if err == sql.ErrNoRows {
			return nex.Errors.DataStore.NotFound
		}

		globals.Logger.Error(err.Error())
		// TODO - Send more specific errors?
		return nex.Errors.DataStore.Unknown
	}

	if underReview {
		return nex.Errors.DataStore.UnderReviewing
	}

	return 0
}
```

So whenever a method needs to interact with an object it always verifies the object is available first. This isn't usable here though since `IsObjectAvailable` checks to see if `upload_completed` is true. This is actually the reason why `GetObjectSizeByDataID` exists, since that method specifically does not call `IsObjectAvailable` and always returns the objects size. `GetObjectInfoByDataID` can't be used in this context to get the objects size because the `upload_completed` check will always fail, due to the upload not being marked as completed yet

Reply:
> So the question becomes, do we add an additional check and helper method named something like `S3RemoveObject` and call that as well, or do we just keep `DeleteObjectByDataID` and assume that the server implementing that function will do it's own S3 clean up

We can let the implementation handle S3, considering the current implementation of `DeleteObject` lets the handler do that

> > We should also check for permissions and if the database doesn't report the upload as complete, but I think that can be done on the delete handler (at least the permissions part)
> 
> Unless I'm misunderstanding the suggestion, this is an implementation detail really and is how things are already implemented by SMM. 

I mean more like in the deletion segment of `CompletePostObject`, since we only call delete if the upload failed, and nothing else. If there aren't any safety checks on the delete handler, I could send a `CompletePostObject` with any data ID and report as upload failed, and the server would delete the data.

Reply:
> I mean more like in the deletion segment of `CompletePostObject`, since we only call delete if the upload failed, and nothing else. If there aren't any safety checks on the delete handler, I could send a `CompletePostObject` with any data ID and report as upload failed, and the server would delete the data.

In the version I have that's not pushed right now I added a helper method to get an objects owner and compare it to the PID of the user making the request, so only an owner of an object can modify it here. But I do think it's a good suggestion to also check if it's been marked as uploaded or not, I'll add that